### PR TITLE
#206 - Add option to convert SparseMatrixExp to explicit matrix

### LINKED ISSUE #206: Add option to convert SparseMatrixExp to explicit matrix
Lazy matrix exponentiation is important for huge matrices to prevent out-of-memory problems. But currently we keep the matrix exponential lazy during the whole analysis, which is a lot slower.

Instead we could make the matrix exponential explicit once in the beginning and continue with the normal analysis.

### PR DESCRIPTION
**Pull request by schillic**
Closes #206.

### CODE REVIEW

**Comment by schillic** [comment]
Unfortunately, for MNA5 this is not a good idea. It takes 223 seconds to make the matrix explicit, and the full matrix multiplication seems to be too expensive (the successor computation would take hours). But at least for FOM we save some seconds.

**Comment by mforets** [comment]
Is the idea to: use the lazy matrix exponential in the discretization + decomposition phases, BUT use the matrix exponential explicitly in the iteration? This option seems OK to me, but the names confusing (`:lazy_expm` and `:make_lazy_expm_explicit`). Maybe add `:lazy_expm_iter`  which defailts to `:lazy_expm`, after `:lazy_expm` has been set / taken its default value.

**Comment by schillic** [comment]
Yes, but actually not even for decomposition. Okay for a different name. We could also merge the two options, but then there are three possible values. Like option `expm` with values `:explicit` (default), `:lazy`, and `:lazy_init`.

**Comment by mforets** [comment]
> Yes, but actually not even for decomposition. 

Ah, OK it is only for discretization of the continuous ODE.

Then what about keeping `:lazy_expm` as it is now, but add `:lazy_expm_discretization` as well.

```
check_aliases_and_add_default_value!(dict, dict_copy, [:lazy_expm], false)

check_aliases_and_add_default_value!(dict, dict_copy, [:lazy_expm_discretization], <if it is not given, default to lazy_expm.>)
```

and make this change in `solve.jl`:


```julia
        Δ = discretize(
            system,
            options[:δ],
            approx_model=options[:approx_model],
            pade_expm=options[:pade_expm],
            lazy_expm=options[:lazy_expm_discretize],
            lazy_sih=options[:lazy_sih]
            )
```

File: src/solve.jl [hunk 1]
```diff
@@ -122,19 +122,46 @@ function solve(system::InitialValueProblem,
...
 
-    # ==============================
-    # Sparse/dense matrix conversion
-    # ==============================
+    # ===================================
+    # Sparse/dense/lazy matrix conversion
+    # ===================================
     A = Δ.s.A
+    create_new_system = false
+    if !options[:lazy_expm] && options[:lazy_expm_discretize]
```
Comment (mforets) [inline src/solve.jl]: you can probably just keep `if !options[:lazy_expm]`

Comment (schillic) [reply src/solve.jl]: No, otherwise the matrix may not be lazy :wink:

Comment (mforets) [reply src/solve.jl]: but then check if the matrix if it is lazy (a `SparseMatrixExp`)

Comment (schillic) [reply src/solve.jl]: too late :snail:

**Pull request merged**